Fixes #31240 - New Host Registration Form

stejskalleos opened the pull request:
* Completely in `React`, with `router` & `PF4` components
* Divided into two tabs: General fields & Advanced Fields
* Slots for extending form fields (general / advanced)
* Form can be submitted only if there are no invalid fields

### Plugin extensions:
* https://github.com/Katello/katello/pull/9249
* https://github.com/theforeman/foreman_remote_execution/pull/574

### Changes
**Organization & Location**
* Can be changed in the Form
* changing ORG/LOC will fetch new data and reset form fields (HostGroup, OS & Smart Proxy)
*  Reset of fields from plugins is handled by plugins.

**Host group**
* If selected, following fields inherit the value from selected host group: Operating System, Activation Keys & Life Cycle Environment
* Selecting HG resets OS field

**Operating System**
* When selected, show info about assigned `host_init_config` template.
* If template is not found, show error and block form

**Smart Proxy** - No changes

**Insecure** - No changes

**Setup Remote Execution** & **Setup Insights**
* `Yes/No` value in default options `Inherit from Host parameter` is delivered from selected organization, location host group and operating system. The chain of getting value is:
`organization -> location -> operatingsystem -> hostgroup`

**Packages** (new field)
* Install packages after the host registration, also: https://github.com/theforeman/foreman/pull/8298

**Repository** & **Repository GPG** (new fields)
* For https://github.com/theforeman/foreman/pull/8402

**Token Life Time**
* Can be set to `unlimited`, otherwise value validated as: `>= 1 && <= 999 999`

stejskalleos committed:
Fixes #31240 - New Host Registration Form

Refactoring of the form, now using React & PF4 components.

ekohl commented on app/controllers/concerns/foreman/controller/registration_commands.rb:
```diff
@@ -29,10 +29,25 @@ def endpoint
 
-    "-H 'Authorization: Bearer #{jwt}'"
+    if registration_params['jwt_expiration'].present?
+      jwt_args[:expiration] = registration_params['jwt_expiration'].to_i.hours.to_i if registration_params['jwt_expiration'] != 'unlimited'
```
Do we want to allow unlimited JWTs? From a security perspective that's IMHO a really bad idea. AFAIK there's no way to revoke a JWT other than rotating the signing key which is a very painful operation. If it somehow leaks (copy-paste in some public channel), it becomes a major hassle.

stejskalleos replied:
Yes we want to, JWT generated for registration are scoped to `RegistrationController` only, so the usage of the token is not unlimited.

ekohl replied:
Please address my comment on how it would be revoked if needed.

stejskalleos replied:
The only way right now is to do it via rails console, as suggested in [BZ](https://bugzilla.redhat.com/show_bug.cgi?id=1919979#c1):
```
foreman-rake console
User.find_by_login('admin').jwt_secret.destroy
```
That way you get invalidated all the tokens generated by the user.
I created [redmine](https://projects.theforeman.org/issues/32278) task for UI implementation, plus update of docs would be also needed (I'll do it when we have new version of Foreman released.)

ekohl replied:
Thank you, that does address my concerns. I somehow missed that the JWT was always tied to a user and assumed we used JWT in a mode where it's signed with a key. This is likely because that's how they're used in the token-based Puppet certificate workflow.

ekohl commented on app/controllers/registration_commands_controller.rb:
```diff
@@ -1,28 +1,59 @@
-    @operating_systems = Operatingsystem.authorized(:view_operatingsystems).select(:id, :title)
-    @smart_proxies = Feature.find_by(name: 'Registration')&.smart_proxies || []
+  def organizations
+    User.current.admin? ? Organization.select(:id, :name) : User.current.my_organizations.select(:id, :name)
```
It feels like `my_organizations` should take care of `User.current.admin?` so it could become:
```suggestion
    User.current.my_organizations.select(:id, :name)
```

ekohl commented on app/views/unattended/provisioning_templates/registration/global_registration.erb:
```diff
@@ -6,6 +6,7 @@ model: ProvisioningTemplate
 #!/bin/sh
 <%
     headers = ["-H 'Authorization: Bearer #{@auth_token}'"]
+    activation_keys = [(@hostgroup.params['kt_activation_keys'] if @hostgroup), @activation_keys].compact.join(',')
```
I kind of feel like `@activation_keys` should already contain this logic.

stejskalleos replied:
`@activation_keys` is already used for keys set by user (from `:activation_keys` URL parameter).
Plus having this in templates helps users to see the logic with host group keys & the keys from user.

ares commented on test/controllers/registration_commands_controller_test.rb:
```diff
@@ -1,19 +1,92 @@
 require 'test_helper'
+require 'pry'
```
leftover

ares commented on app/controllers/registration_commands_controller.rb:
```diff
@@ -1,28 +1,59 @@
+  end
+
+  def locations
+    User.current.admin? ? Location.select(:id, :name) : User.current.my_locations.select(:id, :name)
```
same as above

ares commented on webpack/assets/javascripts/react_app/routes/RegistrationCommands/RegistrationCommandsPage/RegistrationCommandsPageHelpers.js:
```diff
@@ -0,0 +1,92 @@
+  const hostGroupOS = operatingSystems.find(os => `${os.id}` === `${osId}`);
+
+  if (hostGroupOS) {
+    return sprintf('OS from Host group: %s', formatOSname(hostGroupOS));
```
`Host Group` to keep the indentation consistent with the field label

ares commented on webpack/assets/javascripts/react_app/routes/RegistrationCommands/RegistrationCommandsPage/RegistrationCommandsPageHelpers.js:
```diff
@@ -0,0 +1,92 @@
+  if (template?.name) {
+    return (
+      <span>
+        {__('Initial configuration template')}:{' '}
```
This is pretty awesome functionality :-) Not in this PR but could you also add a preview functionality? That could e.g. open a modal window with the rendered version for the current form values. Opening a redmine issue for that would be enough :-)

stejskalleos replied:
https://projects.theforeman.org/issues/32251

ares commented on webpack/assets/javascripts/react_app/routes/RegistrationCommands/RegistrationCommandsPage/components/fields/Repository.js:
```diff
@@ -0,0 +1,81 @@
+      labelIcon={
+        <Popover
+          bodyContent={__(
+            "Repository URL / details, for example for Debian OS family: 'deb http://deb.example.com/ buster 1.0', for Red Hat OS family: 'http://rpm.example.com/'"
```
A repository to be added before the registration is performed. It can be useful to e.g. make the subscription-manager packages available for the purpose of the registration. For Red Hat family distributions, this should be the URL of the repository, e.g. 'http://rpm.example.com/'. For Debian OS families, it's the whole list file content, e.g. 'deb http://deb.example.com/ buster 1.0'.

ares commented on webpack/assets/javascripts/react_app/routes/RegistrationCommands/RegistrationCommandsPage/components/fields/Repository.js:
```diff
@@ -0,0 +1,81 @@
+      label={__('Repository GPG key')}
+      fieldId="reg_smart_proxy"
+      labelIcon={
+        <Popover bodyContent={__('GPG key for the repository')}>
```
"If packages are GPG signed, the public key can be specified here to verify the packages signatures. It needs to be specified in the ascii form with the -----BEGIN PGP PUBLIC KEY BLOCK----- header."

The last sentence is my assumption, also perhaps we can just call it `with the pgp public key header`.

ares commented on webpack/assets/javascripts/react_app/routes/RegistrationCommands/RegistrationCommandsPage/components/fields/SmartProxy.js:
```diff
@@ -0,0 +1,66 @@
+    labelIcon={
+      <Popover
+        bodyContent={__(
+          'Only Smart Proxies with enabled Registration module are available'
```
Only Smart Proxies with Registration feature enabled are displayed.

ares commented on webpack/assets/javascripts/react_app/routes/RegistrationCommands/RegistrationCommandsPage/components/fields/Taxonomies.js:
```diff
@@ -0,0 +1,73 @@
+        isDisabled={isLoading}
+        isRequired
+      >
+        <FormSelectOption value="" label={__('Any Organization')} />
```
Could the value be rather 'Not specified'? Also the inline help could say, "If no Organization is set, the default Organization of the user is assumed."

Likewise for the locations.

ares reviewed (changes requested):
Wow this is way way better! I left few comments regarding texts. Let me know if you need help with the Host Group inline help (currently TODO :-)

There's one styling thing that does not look right, I think the horizontal alignment should be different
![1_misalignment](https://user-images.githubusercontent.com/109773/113151807-c0c52680-9235-11eb-902f-a7435949208a.png)

I'm also attaching the current screenshots in case @https://github.com/theforeman/foreman has any tips
![2_host](https://user-images.githubusercontent.com/109773/113154507-5bbf0000-9238-11eb-82ed-69d621c794f3.png)
![3_host](https://user-images.githubusercontent.com/109773/113154500-5b266980-9238-11eb-8383-07d54ddc8f3a.png)

ares commented on webpack/assets/javascripts/react_app/routes/RegistrationCommands/RegistrationCommandsPage/index.js:
```diff
@@ -0,0 +1,290 @@
+              </Tab>
+              <Tab
+                eventKey={1}
+                title={<TabTitleText>{__('Advanced fields')}</TabTitleText>}
```
Could this be just Advanced and the other just General? The fields is not necessary IMHO

Ron-Lavi commented on config/routes.rb:
```diff
@@ -130,8 +130,10 @@
         get 'preview_host_collection'
 
-        get 'register', to: 'registration_commands#new'
+        match 'register' => 'react#index', :via => [:get]
```
can you use the `get` syntax here as well? same as the changes bellow

Ron-Lavi commented on webpack/assets/javascripts/react_app/routes/RegistrationCommands/constants.js:
```diff
@@ -0,0 +1,6 @@
+export const PATH = '/hosts/register';
```
I think `HOST_REGISTRATION_PATH` might fit more

Ron-Lavi commented on webpack/assets/javascripts/react_app/redux/reducers/index.js:
```diff
@@ -22,6 +22,7 @@ import { reducers as apiReducer } from '../API';
 import { reducers as modelsPageReducers } from '../../routes/Models/ModelsPage';
 import { reducers as settingRecordsReducers } from '../../components/SettingRecords';
 import { reducers as personalAccessTokensReducers } from '../../components/users/PersonalAccessTokens';
+import { reducers as registrationPageReducers } from '../../routes/RegistrationCommands/RegistrationCommandsPage';
```
do we still need those?

Ron-Lavi commented on webpack/assets/javascripts/react_app/routes/RegistrationCommands/RegistrationCommandsPage/RegistrationCommandsPageActions.js:
```diff
@@ -0,0 +1,24 @@
+  REGISTRATION_COMMANDS,
+} from '../constants';
+
+export const dataAction = params => ({
+  key: REGISTRATION_COMMANDS_DATA,
+  url: foremanUrl('/hosts/register/data'),
+  params,
+});
```
you can call the API method right here, e.g:
```js
export const dataAction = params => get({
  key: REGISTRATION_COMMANDS_DATA,
  url: foremanUrl('/hosts/register/data'),
  params,
});
```
and dispatch it from `index.js`, I think it will make more sense

Ron-Lavi commented on webpack/assets/javascripts/react_app/routes/RegistrationCommands/RegistrationCommandsPage/components/fields/Insecure.js:
```diff
@@ -0,0 +1,47 @@
+      isDisabled={isLoading}
+      isChecked={insecure}
+    />
+  </FormGroup>
```
can you align the checkbox a bit? :D

![Screenshot (85)](https://user-images.githubusercontent.com/26363699/113706422-5570ce00-96e7-11eb-87d6-ba9bf3eda0ee.png)

Ron-Lavi commented on webpack/assets/javascripts/react_app/routes/RegistrationCommands/RegistrationCommandsPage/components/fields/Insecure.js:
```diff
@@ -0,0 +1,47 @@
+          <Popover
+            bodyContent={
+              <div>
+                {__(
+                  'If the target machine does not trust the Foreman SSL certificate, the initial connection could be subject to Man in the middle attack. If you accept the risk and do not require the server authenticity to be verified, you can enable insecure argument for the initial curl. Note that all subsequent communication is then properly secured, because the initial request deploys the SSL ...
+                )}
+              </div>
```
text here is kinda long, maybe we can simplify it a bit for the user?

Ron-Lavi commented on webpack/assets/javascripts/react_app/routes/RegistrationCommands/RegistrationCommandsPage/components/fields/TokenLifeTime.js:
```diff
@@ -0,0 +1,96 @@
+        />
+        <InputGroupText>hours</InputGroupText>
+        <InputGroupText>
+          <Checkbox
```
This one also:
![Screenshot (86)](https://user-images.githubusercontent.com/26363699/113706916-e051c880-96e7-11eb-8409-426c77a0e4ee.png)

Ron-Lavi reviewed (commented):
Wow, an amazing improvement from the old PR, great work @stejskalleos!

Ron-Lavi reviewed (commented):
I know that you used the PageLayout component, but here it created big gaps in the page,
A few UX suggestion, but maybe @marisvirik can help more:

![Screenshot (89)](https://user-images.githubusercontent.com/26363699/113708000-602c6280-96e9-11eb-87ba-4993e6f77c83.png)

Ron-Lavi reviewed (commented):
the last thing, I think we should add a good JS integration test so we can catch things before they break on follow-up refactoring

Ron-Lavi commented on webpack/assets/javascripts/react_app/routes/RegistrationCommands/RegistrationCommandsPage/components/fields/OperatingSystem.js:
```diff
@@ -50,7 +49,7 @@ const OperatingSystem = ({
-      dispatch(get(operatingSystemTemplateAction(hostGroupOsId)));
+      dispatch(operatingSystemTemplateAction(hostGroupOsId));
     }
   }, [dispatch, hostGroupId, handleOperatingSystem]);
```
Can you also add `hostGroups` as a `useEffect` dependencies?

Ron-Lavi commented on webpack/assets/javascripts/react_app/routes/RegistrationCommands/RegistrationCommandsPage/components/fields/Repository.js:
```diff
@@ -0,0 +1,85 @@
+            "A repository to be added before the registration is performed. It can be useful to e.g. make the subscription-manager packages available for the purpose of the registration. For Red Hat family distributions, this should be the URL of the repository, e.g. 'http://rpm.example.com/'. For Debian OS families, it's the whole list file content, e.g. 'deb http://deb.example.com/ buster 1.0'. ...
+          )}
+        >
+          <button
+            className="pf-c-form__group-label-help"
+            onClick={e => e.preventDefault()}
+          >
+            <HelpIcon noVerticalAlign />
+          </button>
```
I saw it's being used in many components, can you move it to the `common` dir?

Ron-Lavi reviewed (commented):
overall looks great!! waiting just for the integration test, let me know if you need help with that,

one thing I noticed, is when using the plugins extensions, they might prevent the user from submitting the form, even if the user doesn't really need them,
I think it's OK to have plugins extending the form, though the user should be able to submit it without them, e.g:

 
![Screenshot (91)](https://user-images.githubusercontent.com/26363699/114299536-7d37ab80-9ac4-11eb-932c-9ea56ff542c1.png)

a small design thought some of the `Advanced` fields get collapsed into 3 rows,
there is more free space on the right that we can provide so all rows will be 1 liner, or 2 if needed :)

![Screenshot (92)](https://user-images.githubusercontent.com/26363699/114299673-39917180-9ac5-11eb-88bf-f2e85dcabbce.png)

Big thanks! great work

stejskalleos commented:
Updated, added `<LabelIcon>` component.

> one thing I noticed, is when using the plugins extensions, they might prevent the user from submitting the form, even if the user doesn't really need them,

yes, this is required behavior (from support) for Katello & `activation keys` field

> a small design thought some of the Advanced fields get collapsed into 3 rows,
there is more free space on the right that we can provide so all rows will be 1 liner, or 2 if needed :)

There is no way how in PF4 extend size of label column, or at least I couldn't find it.

stejskalleos committed:
Fixes #31240 - Update Repo & Packages fields

Ron-Lavi reviewed (dismissed):
Thanks @stejskalleos !!
Looks good to me :)

ares commented:
@stejskalleos needs a small rebase now

ares reviewed (changes requested):
I found a problem, when I pick the OS, I see several hints changing and then the OS is unselected again. See the attached gif

![reg1](https://user-images.githubusercontent.com/109773/115256230-5403c480-a12f-11eb-8dfe-9a252944f956.gif)

The same happens when the HG defines the OS and I pick some explicitly. It seems to be related to the fact I picked specific org and location. If I leave them unspecified, it seems to work. However sometimes it works also in a specific scope, just not if HG is selected.

Other than that I didn't find any issue and the form works nicely. Good job!

ares commented:
Oh there's one more thing, not sure if that something we or PF should fix. If I click on the field label, the field doesn't get focus, instead the whole form tab gets focues.
![reg2](https://user-images.githubusercontent.com/109773/115257671-97126780-a130-11eb-8f17-f85cf5767bba.gif)

stejskalleos committed:
Fixes #31240 - Fix issues: Selecting OS, label & field id

stejskalleos commented:
@ares 
both issues fixed, except for selecting `Activation keys` field by clicking on the label, that doesn't work because the component used for the field doesn't support it.

MariSvirik commented:
@stejskalleos just a small nitpick about [PF4 word capitalization](https://www.patternfly.org/v4/ux-writing/capitalization). So PF4 uses sentence case, which means that also your labels should use it (just the first letter of the first word is capitalized) but there are some exceptions e.g. product names etc.
What I spotted was:

Current state - **Correct state in labels**

Host Group - **Host group**
Operating System - **Operating system**
Register Host - **Register host** unless it's also a part of the navigation, then it should Register Host
Install Packages - **Install packages**
Token Life Time - **Token lifetime (hours)**
Smart Proxy - **Smart Proxy** (it's correct)
Activation Keys - **Activation keys**
Setup Insights - **Setup Insights** (this is correct as Insights is a product name)
Remote Execution Interface - **Remote execution interface**

You get the point. 
Anyway, it looks really good. Quite an improvement.

stejskalleos committed:
Fixes #31240 - Fix label names & texts

stejskalleos committed:
Fixes #31240 - Fix label names & texts

ares reviewed (approved):
I've retested the functionality and it feels great! Given the ack from code perspective, ack from the design perspective and my testing, I consider this as ready for merge. Merging this would break Katello tests, so I'll take care of merging the Katello counterpart in sync. I expect this could get in this week so if there's any concern, please speak up!

ares commented:
Thanks @stejskalleos, Katello PR is ready so I'm merging this one! Great work, looking forward to see it at the community demo :-)

tbrisker commented on webpack/assets/javascripts/react_app/routes/RegistrationCommands/RegistrationCommandsPage/components/Actions.js:
```diff
@@ -0,0 +1,59 @@
+  FormHelperText,
+  FormGroup,
+} from '@patternfly/react-core';
+import { ExclamationCircleIcon } from '@patternfly/react-icons/dist/js/icons/exclamation-circle-icon';
```
Looks like this broke the build, this isn't exposed by foreman-js vendor: https://github.com/theforeman/foreman-js/blob/master/packages/vendor-core/lib/modules.js
However I think it should be possible to use 
```
import { ExclamationCircleIcon } from '@patternfly/react-icons';
```
instead

tbrisker replied:
BTW, in the past we had a lint rule that would yell about this, not sure what happened to it.

Ron-Lavi replied:
opened a fix in https://github.com/theforeman/foreman/pull/8467

I also remember a linting rule that used to yell about it,
though maybe it was more for something like `import sortBy from 'lodash/SortBy'`?
Today you can do just `import {sortBy} from lodash` btw :D
